gateway/src/http_entrypoints.rs

```diff
@@ -0,0 +1,357 @@
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+// Copyright 2022 Oxide Computer Company
+
+//! HTTP entrypoint functions for the gateway service
+
+use std::sync::Arc;
+
+use dropshot::{
+    endpoint, ApiDescription, EmptyScanParams, HttpError, HttpResponseOk,
+    HttpResponseUpdatedNoContent, PaginationParams, Path, Query,
+    RequestContext, ResultsPage, TypedBody,
+};
+
+use crate::GatewayService;
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+
+#[derive(Serialize, JsonSchema)]
+struct SpInfo {
+    info: SpIgnitionInfo,
+    details: SpDetails,
+}
+
+#[derive(Serialize, JsonSchema)]
+#[serde(tag = "state")]
+enum SpDetails {
+    Disabled,
+    Unresponsive,
+    Enabled {
+        serial_number: String,
+        // TODO more stuff
+    },
+}
+
+#[derive(Serialize, JsonSchema)]
+#[serde(tag = "present")]
+struct SpIgnitionInfo {
+    id: SpIdentifier,
+    details: SpIgnitionDetails,
+}
+
+#[derive(Serialize, JsonSchema)]
+#[serde(tag = "present")]
+enum SpIgnitionDetails {
+    #[serde(rename = "no")]
+    Absent,
+    #[serde(rename = "yes")]
+    Present {
+        id: u8,
+        power: bool,
+        ctrl_detect_0: bool,
+        ctrl_detect_1: bool,
+        flt_a3: bool,
+        flt_a2: bool,
+        flt_rot: bool,
+        flt_sp: bool,
+    },
+}
+
+#[derive(Serialize, JsonSchema)]
+struct SpComponentInfo;
+
+#[derive(Deserialize, JsonSchema)]
+struct Timeout {
+    timeout: Option<u32>,
+}
+
+#[derive(Serialize, Deserialize)]
+struct TimeoutSelector<T> {
+    last: T,
+    start_time: u64, // TODO
+}
+
+#[derive(Serialize, Deserialize, JsonSchema)]
+#[serde(rename_all = "lowercase")]
+enum SpType {
+    Sled,
+    Power,
+    Switch,
+}
+
+#[derive(Serialize, Deserialize, JsonSchema)]
+struct SpIdentifier {
+    #[serde(rename = "type")]
+    typ: SpType,
+    slot: u32,
+}
+
+type TimeoutPaginationParams<T> = PaginationParams<Timeout, TimeoutSelector<T>>;
+
+#[derive(Deserialize, JsonSchema)]
+struct PathSp {
+    /// ID for the SP that the gateway service translates into the appropriate
+    /// port for communicating with the given SP.
+    #[serde(flatten)]
+    sp: SpIdentifier,
+}
+
+#[derive(Serialize, Deserialize, JsonSchema)]
+struct PathSpComponent {
+    /// ID for the SP that the gateway service translates into the appropriate
+    /// port for communicating with the given SP.
+    #[serde(flatten)]
+    sp: SpIdentifier,
+    /// ID for the component of the SP; this is the internal identifier used by
+    /// the SP itself to identify its components.
+    component: String,
+}
+
+/// List SPs
+///
+/// Since communication with SPs may be unreliable, consumers may specify an
+/// optional timeout to override the default.
+///
+/// This interface may return a page of SPs prior to reaching either the
+/// timeout with the expectation that callers will keep calling this interface
+/// until the terminal page is reached. If the timeout is reached, the final
```

Comment:
What does it mean that the input is paginated? Why not just specify a "timeout duration" as a scalar argument?

Related: Any reason to expose this up, instead of keeping it internal? Or was the choice arbitrary?

Why I'm considering this - does the caller have any idea which values would be "reasonable" for a timeout here? Or is it just a matter of picking something "big enough" and sticking with it?

Reply:
I'm not 100% clear on what you're asking so if I seem to be answering a question you're *not* asking, please redirect me.

The query parameters for paginated endpoints take either the initial inputs **or** the page token.

Why let consumers choose the timeout? I'm not sure it's critical, and I'm not sure how we'd use that. Perhaps with empirical test we can show that this isn't needed. I could imagine consumers potentially having a quick path and a slow path. e.g. if Ignition says it's up but we don't hear from it then maybe we give it some time to see if we can learn anything... or we just reboot it and don't try that hard.